packages/app/src/cli/services/graphql/common.ts

```diff
@@ -46,31 +46,47 @@ export function validateSingleOperation(graphqlOperation: string): void {
 }
 
 /**
- * Validates that the specified API version is supported by the store.
+ * Determines the API version to use based on the user provided version and the available versions.
  * The 'unstable' version is always allowed without validation.
  *
  * @param adminSession - Admin session containing store credentials.
- * @param version - The API version to validate.
- * @throws AbortError if the version is not supported by the store.
+ * @param userSpecifiedVersion - The API version to validate.
+ * @param minimumDefaultVersion - Optional minimum version to use as a fallback when no version is specified.
+ * @throws AbortError if the provided version is not allowed.
  */
-export async function validateApiVersion(
+export async function resolveApiVersion(
   adminSession: {token: string; storeFqdn: string},
-  version: string,
-): Promise<void> {
-  if (version === 'unstable') return
+  userSpecifiedVersion?: string,
+  minimumDefaultVersion?: string,
+): Promise<string> {
+  if (userSpecifiedVersion === 'unstable') return userSpecifiedVersion
 
-  const supportedVersions = await supportedApiVersions(adminSession)
-  if (supportedVersions.includes(version)) return
+  const availableVersions = await fetchApiVersions(adminSession)
 
-  const firstLine = outputContent`Invalid API version: ${version}`.value
-  const secondLine = outputContent`Supported versions: ${supportedVersions.join(', ')}`.value
+  if (!userSpecifiedVersion) {
+    // Return the most recent supported version, or minimumDefaultVersion if specified, whichever is newer.
+    const supportedVersions = availableVersions.filter((version) => version.supported).map((version) => version.handle)
+    if (minimumDefaultVersion) {
+      supportedVersions.push(minimumDefaultVersion)
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    return supportedVersions.sort().reverse()[0]!
```

Comment:
Is it guaranteed that `fetchApiVersions` returns the versions in order?

Reply:
Oh, ignore this. I hadn't seen the `sort()` and I thought you were expecting the backend response to be ordered :man-facepalming: